### ISSUE TITLE
Breaking changes for `StringSet` from 2.5.43 -> 2.5.61

### ISSUE THREAD

**Issue by WeihanLi** [StackExchange/StackExchange.Redis#2071]
The following code breaks when upgrade 2.5.43 -> 2.5.61

``` c#
database.StringSet(_keyName, Base, _expiresIn, When.NotExists);
```

Error message:
```
Argument 4: cannot conv
ert from 'StackExchange.Redis.When' to 'bool'
```

Seemed it's introduced from https://github.com/StackExchange/StackExchange.Redis/pull/2029

**Comment by mgravell** [comment]
Hmmm; indeed, that will be a code-breaking (but thankfully not binary-breaking) change - we must have overlooked that one, sorry. I think we should change `CommandFlags flags` to `CommandFlags flags = CommandFlags.None`, which would allow this previous usage to continue working. Thoughts?

**Comment by WeihanLi** [comment]
Hi @mgravell , thanks for your quick response.

I think maybe it's more simple to adjust the order of the parameters,

For example:

``` diff
- bool StringSet(RedisKey key, RedisValue value, TimeSpan? expiry = null, When when = When.Always, CommandFlags flags = CommandFlags.None)
+ bool StringSet(RedisKey key, RedisValue value, TimeSpan? expiry = null, When when = When.Always, CommandFlags flags = CommandFlags.None, bool keepTtl = false)
```

Another thought is to add a new method such as `StringSetAndKeepTtl`(maybe a better name)

**Comment by mgravell** [comment]
Reordering is a binary break, and we've always tried to maintain a convention of flags as the last param.

**Comment by slorello89** [comment]
Probably needs several smaller overloads, with each ordered sequence of optional parameters. Then in the final version the new parameter should be just prior to the flags.

**Comment by erick2red** [comment]
This caused some pain for me today. While it works perfectly fine in my development setup in Visual Studio, when I deploy to Azure it crashes with the exception below

```
Method not found: 'System.Threading.Tasks.Task`1<Boolean> StackExchange.Redis.IDatabaseAsync.StringSetAsync(StackExchange.Redis.RedisKey, StackExchange.Redis.RedisValue, System.Nullable`1<System.TimeSpan>, Boolean, StackExchange.Redis.When, StackExchange.Redis.CommandFlags)'
```

**Comment by NickCraver** [comment]
@erick2red Interesting - that's _backwards_ - looking for a new method but using an older assembly? It sounds like you're silently (or with a warning) downgrading somewhere vs. what you're compiling against.

**Comment by erick2red** [comment]
I'm compiling against the latest version, and I'm deploying a package to Azure with the binaries in it. I think what's happening is that the runtimes are different, the one in my machine versus the one in the cloud service and the method resolution is tripping there. I fixed it by adding the extra parameters to the call site and making sure it wasn't calling the method with the `bool keepTtl` in it. Somehow, the runtime in Azure can't find that method in the Nuget being uploaded.

**Comment by erick2red** [comment]
Interestingly enough, it doesn't happen with .NET Core 5. IT happens when running .NET Framework 4.7.2.

**Comment by NickCraver** [comment]
I pushed a PR to remedy this in #2098. @mgravell good for eyes when you have a chance - I did try going with the simple `CommandFlags` overload but wanted to add a test suite to be sure and that revealed problems there with ambiguous overloads so...went the adding another overload route. I think this covers all possible source breaks from this.

Sorry about the break here, and thanks for the report!

**Comment by NickCraver** [comment]
This is fixed for the next release (on MyGet now) in #2098 - thank you for the report! We're taken some extra steps to help ensure this doesn't happen going forward.

**Comment by erick2red** [comment]
This is happening again, in `2.6.45`. It's a different method this time tho.
```csharp
        [Browsable(false)]
        [EditorBrowsable(EditorBrowsableState.Never)]
        Task<bool> KeyExpireAsync(RedisKey key, TimeSpan? expiry, CommandFlags flags);

        Task<bool> KeyExpireAsync(RedisKey key, TimeSpan? expiry, ExpireWhen when = ExpireWhen.Always, CommandFlags flags = CommandFlags.None);
```

This is the exception message: 
```
Method not found: 'System.Threading.Tasks.Task`1<Boolean> StackExchange.Redis.IDatabaseAsync.KeyExpireAsync(StackExchange.Redis.RedisKey, System.Nullable`1<System.TimeSpan>, StackExchange.Redis.ExpireWhen, StackExchange.Redis.CommandFlags)'
```

**Comment by NickCraver** [comment]
I'm very curious what your call looks like because that signature looks to be present (tried very hard not to break these). If you can relay your call pattern will look ASAP, and thanks for the report!

**Comment by erick2red** [comment]
These are the two patterns:

```csharp
                    var objectKey = $"{TypesMap[type].TableName.ToLower()}.{key}";
                    var db = ConnectionManager.AccountsRedis.GetDatabase(1);

...

                    var redisTasks = new List<Task>
                    {
                        db.HashSetAsync(objectKey, hashes.ToArray()),
                        db.KeyExpireAsync(objectKey, TimeSpan.FromHours(1)),
                    };
```

and
```csharp
            string objectKey;
            if (IsRelation)
            {
                objectKey = $"{typeof(T).Name.ToLower()}.{String.Join(".", Keys)}";
            }
            else
            {
                objectKey = $"{typeof(T).Name.ToLower()}.{Key}";
            }
...
            db.KeyExpire(objectKey, TimeSpan.FromHours(1));
...
```

**Comment by erick2red** [comment]
So, now what happened here was the same thing as before. For some obscure reason Azure can't seem to find either of these two

```csharp
...
Task<bool> KeyExpireAsync(RedisKey key, TimeSpan? expiry, ExpireWhen when = ExpireWhen.Always, CommandFlags flags = CommandFlags.None);
...
bool KeyExpire(RedisKey key, TimeSpan? expiry, ExpireWhen when = ExpireWhen.Always, CommandFlags flags = CommandFlags.None)
```

I had to go and explicitly specify `flags: CommandFlags.None` to make it work.

**Comment by erick2red** [comment]
So, I found something weird. When Visual Studio deploys to Azure it deploys a zip package, so I went an unzipped the thing and found the StackExchange dll.
Interestingly enough, the version of the dll shown `2.5.61.22961` and something, despite me upgrading to `2.6.45`.

Now, I went ahead and deleted all the build by product folders from under VS, basically `rm -rfv bin obj ...`, and after _rebuild_, it pulls a dll with the correct version.

I need to test now, to clear the `StackExchange.Redis`'s good name, but, the problem is that for some reason, it only breaks in Azure, apparently, VS locally uses the correct version.
